spec/light-client/Lightclient.tla

```diff
@@ -164,66 +179,83 @@ OneStepOfBisection(storedHdrs) ==
         lhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = lh
         rhdr == CHOOSE hdr \in storedHdrs: hdr[1].height = rh
     IN
-    \* pass only the header lhdr[1] and signed header rhdr
-    IF CheckSupport(lh, rh, lhdr[1], rhdr)
+    IF Verify(lhdr[1].NextVP, rhdr) = FALSE
+    THEN <<FALSE, <<(* empty stack *)>> >> \* TERMINATE immediately
+    ELSE \* pass only the header lhdr[1] and signed header rhdr
+      IF CheckSupport(lh, rh, lhdr[1], rhdr)
         (* The header can be trusted, pop the request and return true *)
-    THEN <<TRUE, Tail(requestStack)>>
-    ELSE IF lh + 1 = rh \* sequential verification
-        THEN (*
-             Sequential verification tells us that the header cannot be trusted:
-             If the search request was scheduled as a left branch, then
-             (1) pop the top request (the left branch), and
-             (2) pop the second top request, which is the request for the right branch.
-             Otherwise, pop only the top request.
-             This is the optimization that is introduced in the English spec.
-             One can replace it e.g. with a parallel search.
-             *)
-            LET howManyToPop == IF topReq.isLeft THEN 2 ELSE 1 IN  
-            <<FALSE, SubSeq(requestStack, 1 + howManyToPop, Len(requestStack))>>
+      THEN <<TRUE, Tail(requestStack)>>
+      ELSE IF lh + 1 = rh \* sequential verification
+        THEN (* Sequential verification tells us that the header cannot be trusted. *)
+            <<FALSE, << (* empty stack *)>> >> \* TERMINATE immediately
         ELSE (*
              Dichotomy: schedule search requests for the left and right branches
              (and pop the top element off the stack).
              In contrast to the English spec, these requests are not processed immediately,
-             but one-by-one in a depth-first order (mind the optimization above).
+             but one-by-one in a depth-first order.
              *)
             LET rightReq == [isLeft |-> FALSE, startHeight |-> (lh + rh) \div 2, endHeight |-> rh]
```

Comment:
It might be simpler expressing this logic as a state machine where initial input is end header height to be verified and then state machine have only one pending request at a time and it is trying to verify end header by executing bisection logic. State machine can probably just maintain current header and header to verify (initially half of end header and initial trusted header) and then we update those values upon successful header verification.

Reply:
I believe that the [English spec](https://github.com/tendermint/tendermint/blob/master/docs/spec/consensus/light-client.md) is explicitly against this approach, as the inner call to Bisection is done only if the outer call to Bisection has succeeded. It does not look like we can do that without a stack.